# M01 Homework
 - Name: Sam Remmey
 - Net ID: sqr8ap
 - URL of this file in GitHub: https://github.com/sqr8ap/DS5001-2025-01-R/blob/m01/lessons/M01_Intro/M01_HW.ipynb

In [1]:
import numpy as np
import pandas as pd
import configparser
config = configparser.ConfigParser()

In [3]:
config.read("../../../env.ini")
data_home = config['DEFAULT']['data_home']
output_dir = config['DEFAULT']['output_dir']

In [5]:
!ls -l {data_home}

total 912
drwxr-xr-x@ 8 Samantha  staff     256 Jan 16 08:52 gutenberg
-rw-r--r--@ 1 Samantha  staff  465627 Jan 20 15:27 pg42324.txt


### Load in text file

In [7]:
my_file = f"{data_home}/pg42324.txt"

In [9]:
lines = open(my_file, 'r').readlines()

In [11]:
lines[:5]

['\ufeffThe Project Gutenberg EBook of Frankenstein, by Mary W. Shelley\n',
 '\n',
 'This eBook is for the use of anyone anywhere at no cost and with\n',
 'almost no restrictions whatsoever.  You may copy it, give it away or\n',
 're-use it under the terms of the Project Gutenberg License included\n']

### Convert to a dataframe

In [13]:
text = pd.DataFrame(lines)
text.columns = ['line_str']
text.index.name = 'line_num'

text.head()

line_str
line_num                                                   
0         The Project Gutenberg EBook of Frankenstein, ...
1                                                        \n
2         This eBook is for the use of anyone anywhere a...
3         almost no restrictions whatsoever.  You may co...
4         re-use it under the terms of the Project Guten...

## Question 1.
How many tokens does the raw text have? By raw text, we mean the text as-is, without all of the Gutenberg boilerplate removed.

In [15]:
# Convert lines to tokens

K = text.line_str.str.split(expand=True).stack().to_frame('token_str')
K.index.names = ['line_num','token_num']

In [17]:
len(K)

80985

There are 80985 tokens, or symbol instances.  

## Question 2.
What is the most frequent pronoun in the text?

In [19]:
# To determine the most frequent pronoun, I'll extract a vocabulary of terms

K['term_str'] = K.token_str.str.replace(r'\W+', '', regex=True).str.lower()
K.sample(10)

token_str term_str
line_num token_num                   
5261     9                 a        a
6723     2                me       me
148      11            began    began
1057     1               our      our
5053     6           several  several
6800     7                to       to
7750     1           mission  mission
5445     7             which    which
2819     4               "it       it
6016     7              when     when

In [43]:
V = K.term_str.value_counts().to_frame('n')
V.index.name = 'term_str'
V.head(20)

n
term_str      
the       4575
and       3120
of        2918
i         2918
to        2257
my        1819
a         1497
in        1232
was       1064
that      1060
me         876
with       736
but        711
had        704
you        638
he         625
which      593
it         589
as         573
not        552

The most frequent pronoun in this text is 'i.'

## Question 3.
Which subject pronoun is most frequent in the text we imported in class? 

'She' was the most frequent subject pronoun in the text we imported in class. 

## Question 4.
Provide a brief explanation for this difference, based on what you may know about the two novels.

This difference makes sense given a background on the subjects of these two novels. Persuasion centers around a female protagonist and is told from the third person POV, whereas Frankenstein centers around a male protagonist and is told from the first person POV. Thus, one might expect 'she' to be the most frequent pronoun in Persuasion, and 'I' to be the most frequent pronoun in Frankenstein. Also, on the list of terms from Frankenstein, 'he' shows up before 'she,' which also makes sense given the context. 